### PR TITLE
function calling fixes for together/llama-3.1

### PR DESCRIPTION
**Pull request by kwindla**
Fixes for these function call test cases:

```
test_strings = [
    '<function=hello>{world}</function>',
    '<function=goodbye>{cruel world}"</function>',
    '<function=spacey>{extra space} </function>',
    '<function=noclose>{this has no closing tag}',
    '<function=empty></function>',
]
```

And also extracts a function call from the middle of an output (not just the beginning).

### CODE REVIEW

**Review by filipi87 — APPROVED** [review]
Thanks for fixing this Kwin. 🙌

File: src/pipecat/services/together.py [hunk 1]
```diff
@@ -131,10 +133,24 @@ async def _process_context(self, context: OpenAILLMContext):
...
-                    await self._extract_function_call(context, function_call_accumulator)
+                        text = chunk.choices[0].delta.content
+                        if most_recent_chunk_was_function_call_start_char:
+                            most_recent_chunk_was_function_call_start_char = False
+                            if text == "function":
+                                accumulating_function_call = True
+                                function_call_accumulator = "<function"
+                            else:
+                                await self.push_frame("<" + TextFrame(chunk.choices[0].delta.content))
+                        elif text == '<':
```
Comment (marcus-daily) [inline src/pipecat/services/together.py:144]: This seems to make the assumption that `<` will arrive in its own chunk with nothing else included, as will the word `function` -- just checking if that assumption is valid?

Do the chunks here correspond to the SSE chunks from Together, and if so, is the above documented behaviour, or something we've observed empirically?

File: src/pipecat/services/together.py [hunk 2]
```diff
@@ -164,14 +180,27 @@ async def process_frame(self, frame: Frame, direction: FrameDirection):
...
 
     async def _extract_function_call(self, context, function_call_accumulator):
+        # logger.debug(f"Extracting function call: {function_call_accumulator}")
         context.add_message({"role": "assistant", "content": function_call_accumulator})
 
-        function_regex = r"<function=(\w+)>(.*?)</function>"
+        # Function format regex. Llama 3.1 sometimes adds an extra " or space just before the
+        # </function> tag. This regexp just ignores the extra characters if they are there. (That's
+        # the [\s"]? part of the regex.) Occasionally the </function> close tag is also missing.
+        function_regex = r'<function=(\w+)>(.*?)<\/function>|<function=(\w+)>(.*)'
```
Comment (marcus-daily) [inline src/pipecat/services/together.py:189]: Is the question mark necessary in `.*?` ?

Comment (marcus-daily) [reply src/pipecat/services/together.py:189]: Ah never mind, I assume this is to turn this into a lazy match.

**Comment by LewisWolfgang** [comment]
This looks fine to me, but I will leave it to others to approve, as I have not tested function calling with Together.

File: src/pipecat/services/together.py (hunk 1 above)
Comment (kwindla) [reply src/pipecat/services/together.py:144]: It's not documented. But I did test really heavily and this seems to be the pattern empirically. It makes sense, given what I've seen in general and the Llama 3.1 tokenizer, I think.

My thinking is that we should implement this way, for now, and then revisit when we have a chance to regularize around @chadbailey59's ideas that there should be configurable "structured data" patterns that aren't streamed through the pipeline.

Comment (marcus-daily) [reply src/pipecat/services/together.py:144]: Sounds reasonable!

**Review by marcus-daily — APPROVED** [review]

**Comment by vipyne** [comment]
this seems pretty stale and the issue is fixed.  ok if I close @kwindla ?

**Comment by vipyne** [comment]
closing for now, but feel free to re-open if necessary @kwindla !

**Pull request closed without merging**